### PR TITLE
kv-cache : refactor + add llama_memory_state_i

### PR DESCRIPTION
**Pull request by ggerganov**
cont https://github.com/ggml-org/llama.cpp/pull/13706#issuecomment-2901225909, https://github.com/ggml-org/llama.cpp/pull/13194

Main goal here is to simplify the abstract interface of `struct llama_kv_cache`.

### Overview

Changes to the internal `struct llama_kv_cache` abstract interface:

- Remove `llama_kv_cache::commit()`
- Remove `llama_kv_cache::restore()`
- Remove `llama_kv_cache::sbatch_init()`
- Remove `llama_kv_cache::ubatch_next()`
- Remove `llama_kv_cache::find_slot()`
- Remove `llama_kv_cache_guard`
- Add:

```c++
--- llama-memory.h

    // the interface for managing the memory state during batch processing
    // this interface is implemented per memory type. see:
    //   - llama_kv_cache_unified_state
    //   - llama_kv_cache_unified_iswa_state
    //   ...
    //
    // the only method that can mutate the memory and the memory state is llama_memory_i::apply()
    //
    // TODO: rename to llama_memory_context_i ?
    class llama_memory_state_i {
    public:
        virtual ~llama_memory_state_i() = default;
    
        // consume the current ubatch from the state and proceed to the next one
        // return false if we are done
        virtual bool next() = 0;
    
        // apply the memory state for the current ubatch to the memory object
        // return false on failure
        virtual bool apply() = 0;
    
        // TODO: this might get reworked in the future when refactoring llama_batch
        virtual std::vector<int64_t> & out_ids() = 0;
    
        // get the current ubatch
        virtual const llama_ubatch & get_ubatch() const = 0;
    
        // get the status of the memory state
        virtual llama_memory_status get_status() const = 0;
    };
    
    using llama_memory_state_ptr = std::unique_ptr<llama_memory_state_i>;

--- llama-kv-cache.h

    // split the input batch into a set of ubatches and verify that they can fit into the cache
    // return a state object containing the ubatches and KV cache state required to process them
    // check the llama_memory_state_i::get_status() for the result
    virtual llama_memory_state_ptr init_batch(
            const llama_batch & batch,
            uint32_t n_ubatch,
            bool embd_pooled,
            bool logits_all) = 0;

    // simulate full cache, used for allocating worst-case compute buffers
    virtual llama_memory_state_ptr init_full() = 0;
```

This new interface changes the logic in `llama_decode()` to first make sure that we can fit the input batch into the cache and only after that we start to process the ubatches. This check takes correctly into account SWA masking and also makes sure that the cache will not be modified before we start the actual computation.

*note: the latter is not yet true for the recurrent cache - [see comments in the code](https://github.com/ggml-org/llama.cpp/blob/cf1000143cec4cad6f5f847a7fda46554855c13a/src/llama-kv-cache.cpp#L2309-L2341)*

Another important update in this PR is that the `find_slot()` logic for unified caches is now improved. Before we looked for a slot (i.e. a set of contiguous cells) that is empty in order to place the ubatch in it. We now allow the slot to contain data from the same or other sequence which is masked (either by causality or by SWA):

https://github.com/ggml-org/llama.cpp/blob/2252eefd2b47e408dcb0781b055762de1d71ecee/src/llama-kv-cache.cpp#L574-L621

This change is needed for the next PR, which will optimize the SWA cache to use just `n_swa + n_ubatch` cells and it also has some other nice properties. For example, we no longer have to explicitly prune tokens on successful batch processing, which simplifies the logic significantly and allows us to re-enable speculative decoding for SWA models (will be done also in the next PR).

The worst-graph reserve logic is also refactored and simplified significantly.

There are also some changes to `llama-batch`, but these are mainly to patch things up so that we are able to push the KV cache refactor first. So no need to review the `llama-batch` in deep details - the code there will be reworked soon.

### TODO

- [x] Adapt the recurrent cache to the new interface
- [x] Test optimization workflow

### Next PRs

- [x] Absorb the [batch reduction logic](https://github.com/ggml-org/llama.cpp/blob/051372cbb153140e295bdf5537adc8db6d9174d8/tools/server/server.cpp#L3388-L3443) into `llama_decode`, so that user code does not have to do it (#13845)
- [x] Enable `n_swa + n_ubatch` for SWA cache (#13833)
- [ ] Auto-purge future tokens after successfully processing a batch

### CODE REVIEW

**Comment by ggerganov** [comment]
This PR should not cause any performance changes and the numerical results should be mostly the same (with some small exceptions due to the new logic in `find_slot()`).

Would appreciate some testing and reports for regressions. Thanks.

File: src/llama-kv-cache.h [hunk 1]
```diff
@@ -289,7 +244,7 @@ class llama_kv_cache_unified : public llama_kv_cache {
 
 // utilizes two instances of llama_kv_cache_unified
 //   the first instance is for the non-SWA layers of the model and the second instance is for the SWA layers
-//   upon successful commit, the SWA cache removes old tokens outside the n_swa window
+//   upon successful processing of the batch, the SWA cache removes old tokens outside the n_swa window
```
Comment (ggerganov) [inline src/llama-kv-cache.h:247]:
```suggestion
```

**Comment by ngxson** [comment]
I re-run the ppl test from https://github.com/ggml-org/llama.cpp/pull/13194#issuecomment-2870223856

master at https://github.com/ggml-org/llama.cpp/commit/aa50ba462f63759e1731227f20f5009cfc2a0f16

```
OK:   Final estimate: PPL = 7.8002 +/- 0.17654   ggml-org/gemma-3-4b-it-GGUF:Q4_K_M
OK:   Final estimate: PPL = 37.6848 +/- 1.03389   bartowski/gemma-2-9b-it-GGUF:Q4_K_M
OK:   Final estimate: PPL = 5.9658 +/- 0.11216   lmstudio-community/Phi-3.1-mini-128k-instruct-GGUF:Q4_K_M
OK:   Final estimate: PPL = 5.2653 +/- 0.09581   bartowski/CohereForAI_c4ai-command-a-03-2025-GGUF:IQ1_M
OK:   Final estimate: PPL = 7.3320 +/- 0.16048   unsloth/Llama-4-Scout-17B-16E-Instruct-GGUF:IQ1_S
```

This PR:

```
OK:   Final estimate: PPL = 7.8003 +/- 0.17654   ggml-org/gemma-3-4b-it-GGUF:Q4_K_M
OK:   Final estimate: PPL = 37.6620 +/- 1.03339   bartowski/gemma-2-9b-it-GGUF:Q4_K_M
OK:   Final estimate: PPL = 5.9658 +/- 0.11216   lmstudio-community/Phi-3.1-mini-128k-instruct-GGUF:Q4_K_M
OK:   Final estimate: PPL = 5.2642 +/- 0.09577   bartowski/CohereForAI_c4ai-command-a-03-2025-GGUF:IQ1_M
OK:   Final estimate: PPL = 7.3302 +/- 0.16037   unsloth/Llama-4-Scout-17B-16E-Instruct-GGUF:IQ1_S
```

Some results changed very slightly, so I'm not sure if this is expect

**Comment by ggerganov** [comment]
Yes, I think this difference is expected for SWA models (note Phi currently is disabled SWA, so no difference). It's caused by the different order in which we place the data in memory, due to the `find_slot()` updates. The results become identical with `--swa-full` - can you confirm?

**Comment by ngxson** [comment]
Yes that's right, I added `--swa-full` and now it become identical to `master` version:

```
OK:   Final estimate: PPL = 7.8002 +/- 0.17654   ggml-org/gemma-3-4b-it-GGUF:Q4_K_M
OK:   Final estimate: PPL = 37.7017 +/- 1.03468   bartowski/gemma-2-9b-it-GGUF:Q4_K_M
OK:   Final estimate: PPL = 5.9658 +/- 0.11216   lmstudio-community/Phi-3.1-mini-128k-instruct-GGUF:Q4_K_M
OK:   Final estimate: PPL = 5.2654 +/- 0.09581   bartowski/CohereForAI_c4ai-command-a-03-2025-GGUF:IQ1_M
OK:   Final estimate: PPL = 7.3320 +/- 0.16048   unsloth/Llama-4-Scout-17B-16E-Instruct-GGUF:IQ1_S
```

**Comment by rhvall** [comment]
With the subsequent changes to the `_kv_` structure, are there any substantial changes that would affect how a cached kv is managed?. Along the same line, updates to code in the `examples` (ex. passkey or lookahead)

**Comment by slaren** [comment]
How does this recover from a failed call to `graph_compute`? What is the replacement for `commit`/`restore`?

**Comment by ggerganov** [comment]
> How does this recover from a failed call to `graph_compute`? What is the replacement for `commit`/`restore`?

There are some tricky scenarios in which we could have overwritten some of the data in the cache by the time the error occurs (i.e. processed the first few ubatches, but not all of them yet). Before (i.e. on `master`), we allowed to place ubatches only in empty slots, so we could simply mark the cells back to empty and recover in such cases. But with the new logic, this is no longer guaranteed because we allow to place ubatches in masked slots. This new logic is quite beneficial because it will enable smaller caches for SWA (i.e. `n_swa + n_ubatch` vs `n_swa + n_batch`) and also we don't have to explicitly prune SWA-masked tokens on successful batch, which allows to seamlessly do short rollbacks. The latter is needed for speculative decoding (#13747) and for cases where the last generated chat response can contain a few extra newlines, which are then discarded by the Web UI. In the latter case, if we pruned all tokens strictly by the SWA window (as it is currently on `master`), then this would cause full reprocessing of the context, while with the new logic, we can still rollback and have all necessary cache data available to reuse.

I think that on compute error, the KV cache should be assumed in an undefined state and the application should take necessary steps to recover (i.e. by clearing it and reprocessing the context that is currently needed). Later on, this reprocessing will become seamless, when we start storing the necessary tokens/embeddings information and add the logic for auto-reprocessing whatever is currently missing from the cache.

**Comment by slaren** [comment]
I am mostly concerned about the abort callback functionality. Errors in the backend are likely to be unrecoverable, but I am not sure if the abort functionality makes sense if it leaves the cache in a bad state.

**Comment by ggerganov** [comment]
I admit that I had completely forgotten about the abort callback. Let me see if we can do something about this.

File: src/llama-kv-cache.cpp [hunk 2]
```diff
@@ -293,26 +348,77 @@ llama_pos llama_kv_cache_unified::seq_pos_max(llama_seq_id seq_id) const {
...
+            uint32_t n_ubatch,
+            bool embd_pooled,
+            bool logits_all) {
+    GGML_UNUSED(embd_pooled);
+
+    auto sbatch = llama_sbatch(batch, hparams.n_embd, true, logits_all);
+
+    std::vector<llama_ubatch> ubatches;
+    while (sbatch.n_tokens > 0) {
+        ubatches.push_back(sbatch.split_simple(n_ubatch));
```
Comment (gabe-l-hart) [inline src/llama-kv-cache.cpp:362]: I think I'm reading this right that it's not possible to use `split_equal` with the unified cache after this refactor. If so, I think this will cause problems with the eventual hybrid cache where `split_equal` is required for recurrent child caches (cc @compilade since you [pointed that out to me earlier](https://github.com/ggml-org/llama.cpp/pull/13276#discussion_r2072240777)).

**Review by slaren — COMMENTED** [review]
The solution seems good.

I still think that `n` and `head` should not be part of the KV state. It leads to very confusing code, for example the new `process` function does not receive or do any KV allocation information, so how does it even work? well, it doesn't really work by itself, it depends on the state of the object that must be setup before calling this function. Making functions as close as possible to pure functions that do not depend on any external state greatly improves the code readability.

**Comment by ggerganov** [comment]
> I still think that n and head should not be part of the KV state.

I tried to address this concern in the latest commit. Still need to do the same update for the recurrent cache. Let me know if this seems better now.

**Comment by slaren** [comment]
It's definitely clearer now, but the fundamental problem is still the same, `n_kv` and `head` are still part of the state of `llama_kv_cache_unified`. I think where it would make more sense to have this state is in `llama_memory_state`, but I understand that would require changing a lot of the graph building functions and it is not strictly necessary.

File: src/llama-kv-cache.cpp [hunk 3]
```diff
@@ -438,23 +567,63 @@ bool llama_kv_cache_unified::find_slot(const llama_ubatch & ubatch) {
...
+        for (int s = 0; s < LLAMA_MAX_PARALLEL_SEQUENCES; ++s) {
+            seq_pos_min[s] = cells.seq_pos_min(s);
+        }
+
         bool found = true;
         for (uint32_t i = 0; i < n_tokens; i++) {
-            // TODO: improve to accept cells that are masked by the SWA
-            if (!cells.is_empty(head + i)) {
+            const llama_pos    pos    = ubatch.pos[i];
+            const llama_seq_id seq_id = ubatch.seq_id[i][0];
```
Comment (gabe-l-hart) [inline src/llama-kv-cache.cpp:594]: In local testing trying to use `split_equal` with the unified portion of the hybrid cache, this line is causing problems. Specifically, I think it conflicts with the logic in `llama_sbatch::add_seq_to_ubatch` ([here](https://github.com/ggml-org/llama.cpp/blob/gg/kv-cache-simplify-part3/src/llama-batch.cpp#L90)) where the `ubatch.seq_id` is only populated with a non-null value at position `ubatch.n_seqs`. I'll keep digging to see if there's a simple solution, but wanted to flag this in case it's an easy fix on your end.

Comment (gabe-l-hart) [reply src/llama-kv-cache.cpp:594]: This is probably not the _right_ solution, but this "fixed" the issue on my Granite 4 branch:
```suggestion
            const llama_seq_id seq_id = ubatch.seq_id[i] == nullptr ? ubatch.seq_id[0][0] : ubatch.seq_id[i][0];

```

Comment (compilade) [reply src/llama-kv-cache.cpp:594]: > I'm not quite following the relationship that should be used to extract `seq_id` for the given `pos` element.

Usually, this is relatively simple since the number of tokens per sequence is known in `ubatch.n_seq_tokens`. In fact, here it could probably be possible to use

```suggestion
            const llama_seq_id seq_id = ubatch.seq_id[i / ubatch.n_seq_tokens][0];
```

although there is another approach without divisions, but with nested loops and which would change the indexing for `ubatch.pos[i]` to `ubatch.pos[s * ubatch.n_seq_tokens + j]` where `s` is in `[0, ubatch.n_seqs)` and `j` in `[0, ubatch.n_seq_tokens)`.

File: src/llama-kv-cache.cpp [hunk 4]
```diff
@@ -438,23 +567,63 @@ bool llama_kv_cache_unified::find_slot(const llama_ubatch & ubatch) {
...
+
+            // can we use this cell? either:
+            //  - the cell is empty
+            //  - the cell is occupied only by one sequence:
+            //    - mask causally, if the sequence is the same as the one we are inserting
+            //    - mask SWA, using current max pos for that sequence in the cache
+            //                always insert in the cell with minimum pos
+            bool can_use = cells.is_empty(head_cur + i);
+
+            if (!can_use && cells.seq_count(head_cur + i) == 1) {
```
[the remark this answers is not in the document]
Comment (gabe-l-hart) [reply src/llama-kv-cache.cpp:604]: That diff is gross, but it just adds an extra conditional to the outer check that checks whether `ubatch.n_seqs == 1` and then always uses `ubatch.seq_id[0][0]`.

[the remark this answers is not in the document]
Comment (compilade) [reply src/llama-kv-cache.cpp:604]: @gabe-l-hart It should not be necessary to limit `ubatch.n_seqs` to `1`. This is almost never true for simple splits anyway, except when `n_ubatch` is `1`.

See <https://github.com/ggml-org/llama.cpp/pull/13746#discussion_r2112737519>.

[the remark this answers is not in the document]
Comment (gabe-l-hart) [reply src/llama-kv-cache.cpp:665]: Thanks! Simplifying the multi-sequence-per-token logic would certainly help from a clarity perspective (having recently tried to understand the current implementation and been only partially successful).

File: src/llama-kv-cache.cpp [hunk 5]
```diff
@@ -1620,31 +1813,40 @@ llama_pos llama_kv_cache_unified_iswa::seq_pos_max(llama_seq_id seq_id) const {
...
+    auto sbatch = llama_sbatch(batch, hparams.n_embd, true, logits_all);
 
-    // slide the attention window, forgetting/pruning old tokens that are outside the window
-    if (do_prune) {
-        for (const auto & [seq_id, entry] : pending.pos) {
-            kv_swa->prune_swa(seq_id, entry.pmin, entry.pmax);
-        }
+    std::vector<llama_ubatch> ubatches;
+
+    while (sbatch.n_tokens > 0) {
```
[the remark this answers is not in the document]
Comment (gabe-l-hart) [reply src/llama-kv-cache.cpp:1984]: (for awareness) I've got the simplified non-abstract version implemented in draft against this PR here: https://github.com/ggml-org/llama.cpp/pull/13904

**Review by slaren — APPROVED** [review]
This looks good to me. Some notes for the future:

- `init_batch` and `init_full` could be moved to `llama_memory_i`. There are still many places where `llama_kv_cache` is used directly instead of `llama_memory_i`. `llama_kv_cache_recurrent` probably should not inherit from `llama_kv_cache`, but rather it should be a completely separate implementation of `llama_memory_i`. I believe that's already the plan, and `llama_kv_cache` is only used in this way to simplify the transition, but ultimately functions like `llama_decode` should not depend on `llama_kv_cache`.

- `llama_kv_cache` could probably be renamed to `llama_kv_cache_i` for consistency

- From what I can tell, `llama_kv_cache_unified_state_i`, `llama_kv_cache_unified_iswa_state_i`, `llama_kv_cache_recurrent_state_i` do not need to be interfaces. Is the goal is to hide the implementation details from the header? Since virtual functions have a nonzero performance cost, I would be wary about turning everything into an interface and adding an indirection to every function call, even if it is not likely to have a significant performance impact at the moment.

**Comment by gabe-l-hart** [comment]
Wow, looks like some great progress on [the list](https://github.com/ggml-org/llama.cpp/pull/13746#issuecomment-2921703899) over the weekend! I see that #13904 got closed. I'm _guessing_ that happened automatically when the target branch got deleted? I'll plan to update it on top of the latest changes unless I hear otherwise.

**Comment by ggerganov** [comment]
Yes, it seemed it was closed automatically. Feel free to update it and submit a new PR.

Btw, I think we should first fix the recurrent cache and also make the iSWA cache work with `split_equal`.

**Comment by gabe-l-hart** [comment]
Great, will do. I also saw that #13834 got moved to draft, so wasn't sure the status on that one. With all of the decoupling, I think the recurrent cache should be largely independent of the others (the only change it needs to recurrent is adding the layer filter), so I'm happy to take whatever order makes the most sense on your end.
